### PR TITLE
Wrap up TBD items on circleCI work

### PR DESCRIPTION
**Pull request by bricerisingalgorand**
## Summary

I'm making this PR at the start of work so that I can continuously see that it's running the circleci jobs correctly. Below are the items I am working on:

- [x] Add mac
- [x] Right size instances
- [x] Collect coverage reports
- [x] Make sure parallelism is 1 everywhere
- [x] Maybe break out e2e subs into another job

Will do later
- Implement linter to enforce partition line in test files
- Optimize package installs

### CODE REVIEW

File: testpartitioning/testFiltering.go
Comment (winder) [inline testpartitioning/testFiltering.go:31]: license was added twice

File: testpartitioning/testFiltering.go
Comment (winder) [inline testpartitioning/testFiltering.go:68]: Sorry for this comment, this obviously isn't critical.

Technically the test name isn't going to give you a uniformly distributed partition key, so you would need to hash this:
```suggestion
	h := fnv.New64a()
	h.Write([]byte(s))
	return h.Sum64()
```

This isn't perfect either, since there are tests with the same name in different packages:
```
~$ rg "func Test"| cut -f 2 -d ' '|sort|wc -l
1360

~$ rg "func Test"| cut -f 2 -d ' '|sort|uniq|wc -l
1344
```

Comment (cce) [reply testpartitioning/testFiltering.go:52]: oh interesting, for a more even distribution you could get the filename & line number of the caller too by `runtime.Caller(1)`. Also, passing the `pc` value it returns to `runtime.FuncForPC` could get you a version of the calling function name and package name, here's an example: https://play.golang.org/p/tEW91-qEdqR

Comment (winder) [reply testpartitioning/testFiltering.go:52]: @cce I was thinking about something similar, but wouldn't that give the line/file of the helper function instead of the test function? I guess hashing a random value would work too (unless the partitions should be deterministic).

Comment (cce) [reply testpartitioning/testFiltering.go:52]: `Caller(N)` will go N levels up the stack, so it will get the test function instead of the helper function — Caller() and FuncForPC() are sometimes used in logging/metrics/tracing helper functions to grab names like this for capturing monitoring data.

Deterministic seems good IMO, especially for re-running tests

Comment (cce) [reply testpartitioning/testFiltering.go:52]: I gave a try at implementing @winder's idea along with the `runtime.Caller(1)` in https://github.com/cce/go-algorand/commit/0cf3b2d1a97fba19fb7f225c17487a3b6758f209 CC @algonautshant

File: scripts/travis/before_build.sh
Comment (winder) [inline scripts/travis/before_build.sh:23]: How would we update these binaries when we change libsodium?

File: scripts/travis/build.sh
Comment (winder) [inline scripts/travis/build.sh:47]: It looks like this is only used for windows, what wasn't working? According to the docs windows should be supported.

File: .circleci/config.yml
Comment (winder) [inline .circleci/config.yml:66]: The current version is 1.14.7

File: scripts/travis/build.sh
Comment (bricerisingalgorand) [reply scripts/travis/build.sh:47]: I can't remember the exact error, but when I was having issues with it on circleci's windows machines. I could go ahead and retract this change since we're planning to get windows working after merging these changes

File: scripts/travis/before_build.sh
Comment (bricerisingalgorand) [reply scripts/travis/before_build.sh:23]: Circleci is caching the libsodum builds based on the md5 sums of all files in libsodium fork, so if it ever changes a new build will happen and the cache will get updated

```
    - run:
          name: Get libsodium md5
          command: |
            mkdir -p tmp
            find crypto/libsodium-fork -type f -exec openssl md5 "{}" + > tmp/libsodium.md5
      - save_cache:
          key: 'libsodium-fork-{{ .Environment.CIRCLE_STAGE }}-{{ checksum "tmp/libsodium.md5" }}'
          paths:
            - crypto/libs
```

**Review by algobarb — APPROVED** [review]
I think it'll be good to merge it in soon and start testing the config file with other PRs. There are still a few areas that we can improve on in the next iterations, but this PR is already quite big and does build the tests on Ubuntu Linux and MacOS. I tested this code out on Thursday and it also works as a good base for testing nightly. I also checked that the pipeline run passsed in CircleCI today.

File: Makefile
Comment (tsachiherman) [inline Makefile:26]: I'm not sure that this is even needed. but if it does, the "touch" was somehow disappeared in this refactoring.

**Review by tsachiherman — CHANGES_REQUESTED** [review]
We should not change the `msgp_gen_test.go` files directly, as these are auto-generated files. Instead, we have a project in `algorand/msgp` that does just that. That project would need to be adjusted to generate test files with the proposed changes. Once you've carries out the changes in the msgp library, you'll need to make a "release" for that, and update the release version in the go.mod/go.sum files.

my suggestion : 
1. rollback the changes in the `msgp_gen_test.go`
2. finish with the rest of the changes here.
3. make the changes in the `algorand/msgp` and create a new release.
4. make the msgp changes in the `algorand/go-algorand` to auto-generate the msgp.

File: Makefile
Comment (bricerisingalgorand) [reply Makefile:26]: Thanks I just added it back. It was a while ago, but I think the touch was needed because of some sort of issue with go accessing empty folders on virtualized disks in docker containers.

File: .travis.yml
Comment (algobarb) [inline .travis.yml:16]: Was `AND branch !~ /circleci/` added for testing or is this necessary to add to the travis.yml file?

Comment (bricerisingalgorand) [reply .travis.yml:16]: Oh yea thanks, that was for testing. I'll remove that now

File: .circleci/config.yml
Comment (bricerisingalgorand) [reply .circleci/config.yml]: Thanks! I fixed that

**Comment by bricerisingalgorand** [comment]
> We should not change the `msgp_gen_test.go` files directly, as these are auto-generated files. Instead, we have a project in `algorand/msgp` that does just that. That project would need to be adjusted to generate test files with the proposed changes. Once you've carries out the changes in the msgp library, you'll need to make a "release" for that, and update the release version in the go.mod/go.sum files.
> 
> my suggestion :
> 
> 1. rollback the changes in the `msgp_gen_test.go`
> 2. finish with the rest of the changes here.
> 3. make the changes in the `algorand/msgp` and create a new release.
> 4. make the msgp changes in the `algorand/go-algorand` to auto-generate the msgp.

Thanks! I made the PR to `algorand/msgp` here

File: .circleci/config.yml
Comment (winder) [inline .circleci/config.yml:173]: I think these scripts should be rolled into the CI scripts at `scripts/travis`

There are subtle differences between what you have here and the scripts.

Also, there is a lot of code duplication, for example in the current scripts gotestsum is declared once in the Makefile, now it's installed and called in many places.

File: .circleci/config.yml
Comment (winder) [inline .circleci/config.yml:225]: Call `make integration` so that the `GOTESTCOMMAND` variable is initialized. Also, the go e2e tests wont be captured (missing junit option)

File: .circleci/config.yml
Comment (winder) [inline .circleci/config.yml:191]: Did you try disabling the travis upload to make sure all partitioned coverage files make it to codecov?

**Comment by winder** [comment]
What are the expectations for this PR? Everything looks like it should work. There's no integration with our current CI scripts and lots of copied code, so it still seems like a prototype.

File: daemon/algod/api/server/router_test.go
Comment (winder) [inline daemon/algod/api/server/router_test.go:181]: Do the tests in here need `partitiontest.PartitionTest(s.T())`?

File: daemon/algod/server_test.go
Comment (winder) [inline daemon/algod/server_test.go:28]:
```suggestion
	"github.com/stretchr/testify/require"
	
	"github.com/algorand/go-algorand/test/partitiontest"
)
```

File: daemon/kmd/config/config.go
Comment (winder) [inline daemon/kmd/config/config.go:60]: How was this disabled previously?

File: data/transactions/logic/assembler_test.go
Comment (winder) [inline data/transactions/logic/assembler_test.go:27]:
```suggestion
	"github.com/stretchr/testify/require"

	"github.com/algorand/go-algorand/test/partitiontest"
```

File: test/partitiontest/filtering.go
Comment (winder) [inline test/partitiontest/filtering.go:1]: Rename this to `filtering_test.go` so that it isn't included in production binaries.

File: scripts/travis/build.sh
Comment (winder) [reply scripts/travis/build.sh:47]: I don't really mind the flag, but maybe name it something like `SKIP_GO_INSTALL`

File: .circleci/config.yml
Comment (winder) [reply .circleci/config.yml:225]: Better would be to use update `scripts/travis/integration_test.sh`

File: test/scripts/e2e.sh
Comment (winder) [inline test/scripts/e2e.sh:131]: Choose a `SKIP` or `ENABLE` pattern, not both. I think `ENABLE_E2E_SUBS`, `ENABLE_E2E_GO` and `ENABLE_E2E_EXPECT` would be easier for someone new to these scripts to understand, `SKIP_*` might be easier to integrate.

File: .circleci/config.yml
Comment (winder) [reply .circleci/config.yml:225]: Maybe `NO_BUILD=true` would help? I saw that flag below, not sure when it helps.

File: test/scripts/e2e_go_tests.sh
Comment (winder) [inline test/scripts/e2e_go_tests.sh:94]: Why use circleci tests split instead of the partitioning?

**Comment by algobarb** [comment]
> What are the expectations for this PR? Everything looks like it should work. There's no integration with our current CI scripts and lots of copied code, so it still seems like a prototype.
> 
> This would be significantly easier to review if the test partitioning changes were made separately. Could add those in a separate PR so it's easier to see the scripting changes?

I think one of the expectation is to merge this PR into master so that we can start getting more branches running CircleCI. This is our last week on our extended trial with Circle CI, so we want to get an idea of what regular CircleCI usage would look like in terms of costs and regular usage. Since this PR absorbed the hackathon work, it will be good if it can be merged so that others can submit smaller PRs afterwards to improve on the script. It is a lot of changed files at once though...

File: .circleci/config.yml
Comment (bricerisingalgorand) [reply .circleci/config.yml:191]: oh good call, I just checked that circleci was showing up here https://codecov.io/gh/algorand/go-algorand/commit/00f33c0090e99b2f00c6ab36a6087c50cb7b9bd3/build

I'll test that too

File: daemon/kmd/config/config.go
Comment (bricerisingalgorand) [reply daemon/kmd/config/config.go:60]: I didn't see a way to disable them before. We only had to do this for circle because they don't allow you to scan usbs on the osx agents

File: test/scripts/e2e_go_tests.sh
Comment (bricerisingalgorand) [reply test/scripts/e2e_go_tests.sh:94]: I'm not actually sure what this change is, could one of you two explain? @algonautshant @cce

File: daemon/kmd/config/config.go
Comment (winder) [reply daemon/kmd/config/config.go:60]: If you needed this, I guess it means kmd crashes if HID is not supported. We should probably fix that instead of disabling KMD on circle. Or at least create an issue for fixing the issue and removing this workaroud.

File: test/scripts/e2e_go_tests.sh
Comment (cce) [reply test/scripts/e2e_go_tests.sh:94]: oh, this is from when I was evaluating the "circleci tests split" runner before Shant's partitioning stuff was ready. you can remove these lines to get rid of it .. as you noticed the `elif [ "${#TESTPATTERNS[@]}" -eq 0 ]; then` is the regular code path we have used and the partitioning stuff is now using.

Comment (winder) [reply test/scripts/e2e_go_tests.sh:94]: You would still need to keep the `gotestsum` call, or set `GOTESTCOMMAND` to include the junit stuff

File: daemon/kmd/config/config.go
Comment (cce) [reply daemon/kmd/config/config.go:60]: While digging into it with @bricerisingalgorand and it seems like without forking the https://github.com/karalabe/hid project and modifying it to use a new fork of the https://github.com/signal11/hidapi project it would be hard to fix the issue, the crashes are happening deep in some C code embedded in karalabe/hid. Some related tickets
https://github.com/signal11/hidapi/issues/205
https://github.com/signal11/hidapi/issues/337

In general, the ability to disable USB probing for a ledger via configuration seemed like a useful feature anyway, and we haven't seen this crash from users, so it seemed like a pretty specific corner case related to the way the CI VMs are set up.

Comment (winder) [reply daemon/kmd/config/config.go:60]: yikes, too bad. I was hoping it would be a matter of checking for this error:
```
func LedgerEnumerate() ([]hid.DeviceInfo, error) {
	if !hid.Supported() {
		return nil, fmt.Errorf("HID not supported")
	}
```

File: daemon/algod/api/server/router_test.go
Comment (cce) [reply daemon/algod/api/server/router_test.go:181]: I think just this TestTestSuite function would need it, the other functions are all called from this test via the testify/suite feature

File: .circleci/config.yml
Comment (cce) [reply .circleci/config.yml:173]: It seems hard to maintain scripts that handle both Travis and Circle runs, especially since we are basically trying to cut up what Travis does. Re: code duplication, that is a matter of personal preference, but I will point out that if later we remove Travis automation, there will be less duplication, and a lot of this can migrate to CircleCI-specific scripts. However IMO total DRY-proofing this PR seems like a big request given the intermediate state we want to be in running Travis and Circle side-by-side.

File: .circleci/config.yml
Comment (cce) [reply .circleci/config.yml:225]: For some background, I was purposefully trying to avoid updating any of the Travis scripts when I worked on this, so that they could run side-by-side without disrupting the Travis system.

**Comment by cce** [comment]
> There's no integration with our current CI scripts and lots of copied code ...

I would disagree with this characterization. Most of the shell snippets in the `run` sections are just setting up environment variables to call the current CI scripts — configure_dev.sh, install_buildtools.sh, run_integration_tests.sh. The only copied code besides environment setup IMO is the line where it calls gotestsum directly given a list of packages.

File: test/scripts/e2e_go_tests.sh
Comment (cce) [reply test/scripts/e2e_go_tests.sh:94]: the entire `if USE_CIRCLECI_SPLIT` section can be removed. This file doesn't need to be changed for this PR.

File: .circleci/config.yml
Comment (winder) [reply .circleci/config.yml:173]: I disagree, these scripts are already platform agnostic. We've been gradually removing the travis specific aspects for a long time. Most of the heavy lifting happens in the Makefile anyway.

For example, I think this section could be replaced with:
```
            export KMD_NOUSB=True
            export PARTITION_TOTAL=$CIRCLE_NODE_TOTAL
            export PARTITION_ID=$CIRCLE_NODE_INDEX
            scripts/travis/test.sh
```

To take it one more step you could set those 3 environment variables in `test.sh`, I don't think that part is as important.

Comment (cce) [reply .circleci/config.yml:173]: Moving all the environment things into "if CIRCLECI" sections is definitely something you could do to bring the remaining stuff into the existing scripts.. I haven't tested whether Makefile dependency resolution works with the kind of workspace-passing between jobs/VMs that Circle and Github Actions provide though, because file timestamps would change, that seems potentially tricky.

File: daemon/kmd/config/config.go
Comment (cce) [reply daemon/kmd/config/config.go:60]: sadly supported is just a `return true` https://github.com/karalabe/hid/blob/master/hid_enabled.go#L61-L66 based on whether you have the right platform-specific build tag enabled or not. the crash was coming from hid.Enumerate

File: test/scripts/e2e_go_tests.sh
Comment (cce) [reply test/scripts/e2e_go_tests.sh:94]: @bricerisingalgorand could you please apply this to your branch? https://github.com/bricerisingalgorand/go-algorand/commit/5a1b71eec5b229ba906442bafce8fcc5cbc3e330

File: test/scripts/e2e_go_tests.sh
Comment (winder) [inline test/scripts/e2e_go_tests.sh]: Unfortunately, I don't think this change is quite right. `GOTESTCOMMAND` isn't being set and it defaults to `go test`, so after this change we'll revert back to `go test`.  If we switch to calling `make integration` instead of no longer be setting `./test/scripts/run_integration_tests.sh`, then `GOTESTCOMMAND` will be set but wont include `--junitfile $TEST_RESULTS/results.xml --jsonfile $TEST_RESULTS/testresults.json`.

Comment (cce) [reply test/scripts/e2e_go_tests.sh]: This is some dead code and so this code path was not being exercised because USE_CIRCLECI_SPLIT is not set anywhere. This commit should have no effect on the tests we're running in CircleCI.

I'm not sure how running make targets will work after attaching a workspace saved by a different job.

**Review by algobarb — APPROVED** [review]

**Review by cce — APPROVED** [review]

**Comment by bricerisingalgorand** [comment]
Could you update your review if you think it looks ok @tsachiherman?

**Review by winder — APPROVED** [review]
Approved pending follow-up tickets (my priorities in parens):
1) verify codecov report still working and correct (high)
2) router_test.go test suite isn't partitioned (high)
3) unify gotestsum configuration (medium)
4) consistent SKIP_E2E_SUBS / E2E_SUBS_ONLY convention - `SKIP_` or `ENABLE_`, not both. (medium)
5) `NO_GIMME` -> `SKIP_GO_INSTALLATION` (medium)
6) verify that go e2e tests are included in reporting, I think that gotestsum may not be properly initialized (medium)
7) unify script changes (medium-low)
8) fix import orders to match our convention (low)

**Pull request merged**
